### DISCUSSION TITLE
Welcome to grblHAL Discussions!

### DISCUSSION THREAD

**Discussion by terjeio** [terjeio/grblHAL#129] [Ideas]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by phil-barrett** [comment]
Hello.  Phil Barrett, designer of the [grblHAL Teensy 4.1 Breakout Board](https://github.com/phil-barrett/grblHAL-teensy-4.x). 

I also run the website at grbl.org. I would be happy to have suggestions (and collaborators) on that!

I am so happy to see this new discussions feature.  Issues has been the channel for discussions up to now and this new feature is far more conducive to that.

**Comment by mundsen** [comment]
Hi
I have ordered the grblHAL Teensy 4.1 from Phil Barrett
Is it possible to combine this with an LCD display/control panel using i2c or serial interface?

I plan to test this controller for my Laser (my own design) in combination with an laser module from Opt Lasers

![3920d2c1-ce67-4673-988c-3fb9de9d2b2c](https://user-images.githubusercontent.com/3058577/101670088-0aa1bb00-3a53-11eb-8d6e-4f98f75f1168.PNG)

https://optlasers.com/kits-and-adapters/plh3d-6w-xf-universal-kit-for-cnc-machines

And LightBurn as Software

This is my laser V2

In laser V1 I used the MiniGerbil controller - but it only support two axis - for my V2 I need 4 axis (Z + rotation)

**Comment by phil-barrett** [comment]
Welcome, Oyvind.  For those who might be interested, Lightburn works ok at Compatibility Level 0 (ie, default build level). They don't support $14 for control pin input inversion so that would need to be done manually or via another sender that does.

**Comment by terjeio** [comment]
@mundsen Welcome, nice machine! Good to see another Norwegian here.

> Is it possible to combine this with an LCD display/control panel using i2c or serial interface?

grblHAL has an open architecture that is highly hackable without changing much of the existing code. Often a single call to an init function from the driver (driver.c) will be enough to expand functionality, and not even that if it is [for private use](https://github.com/terjeio/grblHAL/tree/master/templates).

For my DIY CO2 laser I have added buttons and a small OLED display connected via I2C, keypad via the [keypad plugin](https://github.com/terjeio/grblHAL/tree/master/plugins/keypad) and the OLED display via custom code "hacked" into grblHAL and driver code for the display in the [keypad microcontroller](https://github.com/terjeio/I2C-interface-for-4x4-keyboard).
It has buttons for jogging, homing, testing the laser and adjusting the power level. The display is simple: just for power level, coolant temperature, lid open and jog mode. 

For my lathe I use both the serial interface, I2C interface and controller signal pins connected to my DIY [MPG/DRO](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack).  When not in control of the lathe it just listen passively to the serial output and [parses the real-time commands](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/blob/master/Firmware/MPG%20and%20DRO%20processor/grblcomm.c) for information to keep the LCD updated.

These are just two examples of what is possible, however I am not too keen to add support for all kinds of LCD panels and/or control panels as standard to grblHAL myself. It would be great if someone else would provide code, and I will happily help out with how to "hack" it into grblHAL.

**Reply by mjscosta** [reply]
Hi @terjeio, on my opinion keeping grbl free from LCD/or other displays will be a great conservative approach on keeping grbl simple and maintainable, that's one thing where Marlin (https://github.com/MarlinFirmware/Marlin), failed greatly ...

On my opinion, having a standalone board to manage the LCD, maybe SD Card its the best way to go and more modular, and avoiding having the display code introduce issues with the grbl timing ...

Though one nice feature to have, is to be able to have is to allow grbl to run jobs from the sd card, providing read and write commands to send gc code files via USB or another interface to the SD and later send a command to start the job.

The reason for that is to allow the grbl to run at higher speeds and accelerations avoiding possible issues with bottlenecks on the USB or serial channel or even issues with the host PC sending gcode at different speeds due to different loads, also it would make it more robust to failures in the host pc. Not sure if this is already supported.

**Comment by terjeio** [comment]
> Though one nice feature to have, is to be able to have is to allow grbl to run jobs from the sd card

Many drivers/boards already has support for this via the SD card [plugin](https://github.com/terjeio/grblHAL/tree/master/plugins).
Uploading is not generally supported, could be added by streaming to the SD card instead of executing the code? A new $-command to enable and then run the program to upload it? This would perhaps be compatible with most senders...

**Reply by mjscosta** [reply]
Yeah, there as some of those GCodes in Marlin, extending to upload and stream to the sd card via gcode sender, also to list the sd card and a couple more functions. Though there was an issue with the SD card implementation and support for the SKR V1.1, what caused it was fixed pins in the sd card library, or lack of support for second SPI port or something like that, the details are not relevant, there was a branch of someone working on that for Marlin but I didn't had the time to invest in it.

Having specific support for each mcu driver, can end up being a big effort ... I'll check whats available.

**Comment by mundsen** [comment]
Hi, is there an alternetive to GRBLHAL Teensy with on board stepper drivers like TCM2130..?

I like the Teensy based controller, but would like a more compact setup for my DIY laser project

I need 4 axis - X,Y, Z and a rotary axis + the ability to control a solenoid valve for my air assist

I`m testing different controllers trying to find a controller.. Ilike in combination with LightBurn.

At the moment I`m using an old Smoothieboard5x   

There is a lot of alternatives, like GRBL_Hal, GRBL_ESP32, GRBL_LPC, Smoothieware..

Why should I use GRBL_HAL when compared to the others?

**Reply by terjeio** [reply]
> Hi, is there an alternetive to GRBLHAL Teensy with on board stepper drivers like TCM2130..?

I am working on Trinamic support for grblHAL and have a Bigtreetech SKR v1.3 board currently on my desk with a TMC5160 driver for testing. The 1.4 board should be usable too - to be tested when I am satisfied with the plugin. TMC2130 will be supported too. 

The next iteration of my [Nucleo-64 breakout](https://github.com/terjeio/CNC_Breakout_Nucleo64) is another alternative when verified. The design is completed, I need to order a batch of cards for test. This board will support the [laser plugins](https://github.com/terjeio/grblHAL/tree/test/plugins/laser) and has optocoupled (optionally isolated too) for I/O.

There may be other boards that can be used too - but need a bit of glue code for Trinamic support.

> Why should I use GRBL_HAL when compared to the others?

Good question, no good answer? Active development, new feature, faster processors, plugin support... For high speed laser engraving top of the pack, at least with a Teensy based controller? See issues #142 and #95.

**Comment by mundsen** [comment]
Cool :-)

I have the grblHAL Teensy 4.1, but I need a lot of extra stuff, including 4 external stepper drivers.. to get it up and running - butI`ll consider doing that if the Teensy is the way for high speed..

I have ordered some TMC2130 stepper drivers - do you plan to support SPI for the drivers or step/dir only..?

I have two  Leadshine DM422S from aliexpress, but I do not know if they are fake - can`t find a user guide..

An I have a Makerbase MKS SGen_L V2.0  with a 120Mhz LPC1768 - I do not know if it can be used? How is LPC compared to Teensy 4.1 ?

The Nucleo based controller sounds interesting - long before it is ready for testing? I may help?

I like the active development part :-)

**Reply by terjeio** [reply]
> do you plan to support SPI for the drivers or step/dir only

The plugin code is for SPI support.

> An I have a Makerbase MKS SGen_L V2.0 with a 120Mhz LPC1768 - I do not know if it can be used?

I think so, there is a map file for that in the test branch? [mks_sbase_map.h](https://github.com/terjeio/grblHAL/blob/test/drivers/LPC1769/src/mks_sbase_map.h) I have not tested this as I do not have a board available.

>  How is LPC compared to Teensy 4.1 ?

Slow? No FPU, 120MHz vs 600MHz... E.g. the stepper driver interrupt latency is 2.2µs vs 0.3µs. Still a lot faster than 8-bit Grbl.

> The Nucleo based controller sounds interesting - long before it is ready for testing? 

Hopefully not - my number one priority is to replace the master branch with the test branch, after that is done?

>  I may help?

Maybe, do you have a good understanding of how Trinamic drivers work?
